### DISCUSSION TITLE
Register: Offset based on the length (xxx32 = -2, xxx16 = -1)?

### DISCUSSION THREAD

**Discussion by gutschein** [aldas/modbus-tcp-client#91] [Q&A] [answered]
It seems to me, that some Modbus device doesn't need the register addresse from the manufactors doc 1:1, but with an offset depending on the length. e.g. 40068 is int32 must be requested with an offset -2: 40066 provides proper results. Another register 40096 is int16 must be requested with an offset -1: 40095 for values

I don't know why some manufactors don't calculate the offset in the doc already for there customers. makes no sense from my perspective as the length is fixed for each register.

Before I implement such an offset compensation based on the type length the question: Do you have heard of this? Or is this a rookies question and this is usual? And you may have implemented this as an option already? I suppose no, but I want to sure before I start.

**Comment by aldas** [comment, marked as answer]
Starting address and offset is device/manufactures specific - how they choose to address data. In Modbus packet addresses are 0-based - so first address is 0 (term `data address`). In documentation they are sometimes/often referred as 1-based.  Often in form of `40001` where first number represents function type because some modbus devices can have different memory for different functions - so FC3 can not read FC1  data.

If you want to read more see modbus spec page 7 - "4.4 MODBUS Addressing model"
https://www.modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf
![image](https://user-images.githubusercontent.com/2320301/120267229-3720ec00-c2ac-11eb-9acc-51b04ad097d6.png)


See this https://github.com/aldas/modbus-tcp-client/issues/79#issuecomment-813558442

I highly recommend:
* https://www.simplymodbus.ca/FAQ.htm FAQ section has helped me over the time a lot
* https://modbus.org/docs/Modbus_Application_Protocol_V1_1b3.pdf  reading modbus own spec documentation first parts is also helpful to understand how it work on low level.
* https://www.se.com/se/en/faqs/FA221252/ about addressing
* https://elkortechnologies.com/support/knowledgebase.php?article=14 about addresing

**Comment by gutschein** [comment]
Thank you, Martti.
As I like to have a generic solution and the devices handle it different I implemented an optional offset with an constant part and/or a typ-based part. By now it seems to cover the different need of different manufactors. Just a few code lines. I just wanted to be sure that it isn't double implemented in your library and my wrapper.

Topic is solved for me. :)
